### DISCUSSION TITLE
Passing an argument from Model.prototype.METHOD to Schema.prototype.pre() methods?

### DISCUSSION THREAD

**Discussion by samislam** [Automattic/mongoose#12308] [Q&A]
I need to pass an argument from the `Model.prototype.create()` method to the `Schema.prototype.pre('validate')` method somehow?
Is there any way I can do it?
I need for example to do that in multiple places, I need for instance to pass the express request object to Mongoose, I.e.: passing an argument _from the controller to the model_. 

My case: 

```js
TransactionModel.create({
  from: req.body.from, // id (user object id)
  to: req.body.to, // id (user object id)
  amount: req.body.amount // number, ex: 20$
})
```

```js
const TransactionSchema = new mongoose.Schema({
from: new mongoose.Schema({ name: String, age: Number, balance: Number }),
to: new mongoose.Schema({ name: String, age: Number, balance: Number }),
amount: Number // the amount of money to transfer
})
```
```js
TransactionSchema.pre('save', function(){
   const doc = this
   const fromDoc = await UserMode.findById(doc.from) // to save the user document itself instead of the reference id
   const toDoc = await UserMode.findById(doc.to) // to save the user document itself instead of the reference id
  fromDoc.deductBalance(doc.amount)
  toDoc.feedBalance(doc.amount)
})
```

The problem is that Mongoose is complaining that **_Casting to Embedded failed_**.

How to pass an argument from my express controllers to Mongoose (I call it the model because I use MVC) without validations?

*** for reference: here's the UserModel:
```js
const UserSchema = new mongoose.Schema({
  name: String,
  age: Number,
  balance: Number,
  email: {
     type: String,
     required: [true, "a user must have an email address"]
  }
})

UserSchema.methods.deductBalance = async function (amount){
  const doc = this
  doc.balance = doc.balance -  amount;
  return await doc.save()
}

UserSchema.methods.feedBalance = async function (amount){
  const doc = this
  doc.balance = doc.balance +  amount;
  return await doc.save()
}
```

**Comment by vkarpov15** [comment]
Set the values you want to pass on `$locals`, that's where you can store any extra data you want associated with your document that you don't want stored in MongoDB.

```javascript
const t = new TransactionModel({
  amount: req.body.amount // number, ex: 20$
})
t.$locals.from = req.body.from; // id (user object id)
t.$locals.to = req.body.to; // id (user object id)
await t.save();

// In pre()
TransactionSchema.pre('save', function(){
   const doc = this
   const fromDoc = await UserMode.findById(doc.$locals.from) // to save the user document itself instead of the reference id
   const toDoc = await UserMode.findById(doc.$locals.to) // to save the user document itself instead of the reference id
  fromDoc.deductBalance(doc.amount)
  toDoc.feedBalance(doc.amount)
})
```

**Reply by samislam** [reply]
Good. Can I use this concept to translate the Mongoose custom validation error messages as well?
For instance:
```js
const xSchema = new mongoose.Schema({
age: {
  type: String,
  required: [true, req.body.lang === "en"? "Age is required" : "يرجى ادخال العمر"],
}
})
```
How I can translate these messages if the 
Desired language is determined in the request object?

**Reply by vkarpov15** [reply]
You can set `lang` on the document's `$locals`. Or just the entire `req`:

```javascript
const t = new TransactionModel({
  amount: req.body.amount // number, ex: 20$
})
t.$locals.req = req;
await t.save();
```

**Reply by samislam** [reply]
And how to translate the validation message afterwards? Using middlewares?

**Reply by vkarpov15** [reply]
@samislam yes, currently you would have to use error handling middleware. `xSchema.post('save', function(err, res, next) { if (err instanceof mongoose.Error.ValidationError) { ... } })`. We'll add an issue to allow transforming error message based on document
